src/Stratis.Bitcoin/Consensus/ConsensusRuleEngine.cs

```diff
@@ -262,16 +268,18 @@ private void HandleConsensusError(ValidationContext validationContext)
             this.logger.LogTrace("(-)");
         }
 
-        private void ExecuteRules(IEnumerable<SyncConsensusRule> rules, RuleContext ruleContext)
+        private void ExecuteRules(IEnumerable<SyncConsensusRule> rules, RuleContext ruleContext, RuleType rulesType)
         {
             try
             {
-                using (new StopwatchDisposable(o => this.PerformanceCounter.AddBlockProcessingTime(o)))
-                {
-                    ruleContext.SkipValidation = ruleContext.ValidationContext.ChainedHeaderToValidate.IsAssumedValid;
+                ruleContext.SkipValidation = ruleContext.ValidationContext.ChainedHeaderToValidate.IsAssumedValid;
 
-                    foreach (SyncConsensusRule rule in rules)
+                foreach (SyncConsensusRule rule in rules)
+                {
+                    using (this.performanceCounter.MeasureRuleExecutionTime(rule, rulesType))
```

Comment:
Is it a good idea to lock the counter on each rule iteration?

You could collect samples and at the end push all samples to the counter inside one lock.

Reply:
It locks at the end of each dispose which is executed on each rule.